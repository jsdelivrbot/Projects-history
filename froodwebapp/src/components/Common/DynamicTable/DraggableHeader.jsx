/* eslint-disable
   babel/new-cap,
   react/no-array-index-key,
   react/no-find-dom-node,
   no-param-reassign
*/
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { findDOMNode } from 'react-dom';
import { DragSource, DropTarget } from 'react-dnd';
import { SortIcon } from 'icons';
import styles from './StickyTable.scss';

const headerSource = {
  beginDrag: props => ({
    id: props.id,
    index: props.index
  })
};

const headerTarget = {
  hover: (props, monitor, component) => {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.left - hoverBoundingRect.right) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the right
    const hoverClientX = clientOffset.x - hoverBoundingRect.right;

    // Only perform the move when the mouse has crossed half of the items width
    // When dragging left, only move when the cursor is left 50%
    // When dragging right, only move when the cursor is right 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    // time to update headers order
    props.moveHeader(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex;
  },
  drop: (props, monitor) => {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    props.saveHeaderOrder(dragIndex, hoverIndex);
  }
};

const sourceCollect = (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging()
});

const targetCollect = connect => ({
  connectDropTarget: connect.dropTarget()
});

export class DraggableHeader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortOrder: 'ASC'
    };
  }

  handleSortByColumn = () => {
    this.setState({
      sortOrder: this.state.sortOrder === 'ASC' ? 'DESC' : 'ASC'
    },
    this.props.handleSort(this.props.headerId, this.state.sortOrder)
    );
  }

  render() {
    const {
      headerText,
      headerDataType,
      index,
      isSortable,
      connectDragSource,
      connectDropTarget,
      isDragging
    } = this.props;

    const opacity = isDragging ? 0 : 1;

    return connectDragSource(connectDropTarget(
      <div
        id={ `sticky-header-cell-${index + 1}` }
        className={ styles.cell }
        key={ headerText }
        style={ {
          opacity,
          cursor: 'move',
          textAlign: headerDataType.includes('decimal') ? 'right' : 'left'
        } }
      >
        { headerText }
        { isSortable &&
          <SortIcon
            onClick={ this.handleSortByColumn }
          />
        }
      </div>
    ));
  }
}

DraggableHeader.propTypes = {
  headerId: PropTypes.number.isRequired,
  headerText: PropTypes.string.isRequired,
  headerDataType: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
  isDragging: PropTypes.bool.isRequired,
  isSortable: PropTypes.bool.isRequired,
  handleSort: PropTypes.func.isRequired,
  connectDragSource: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired
};

const source = DragSource('cell', headerSource, sourceCollect)(DraggableHeader);
const DragHeader = DropTarget('cell', headerTarget, targetCollect)(source);

export default DragHeader;
