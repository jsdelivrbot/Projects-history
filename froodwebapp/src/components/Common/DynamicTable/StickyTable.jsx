/* eslint-disable react/no-array-index-key, react/sort-comp */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import elementResizeEvent from 'element-resize-event';
import { Table, Row } from './TableComponents';
import styles from './StickyTable.scss';

const getColRowCount = rows => ({
  rowCount: rows.length,
  columnCount: (rows[0] && React.Children.toArray(rows[0].props.children).length) || 0
});

class StickyTable extends Component {
  constructor(props) {
    super(props);
    const rc = getColRowCount(props.children);

    this.suppressScroll = false;

    this.state = {
      rowCount: rc.rowCount,
      columnCount: rc.columnCount,
      stickyColumnCount: props.stickyColumnCount === 0 ? 0 : (props.stickyColumnCount || 1),
      stickyHeaderCount: props.stickyHeaderCount === 0 ? 0 : (props.stickyHeaderCount || 1)
    };
    this.maxHeight = this.getNodeSize(document.getElementById('main')).height * 0.65; // set max height at 83% of main container
  }

  componentDidMount() {
    this.table = document.getElementById('frood-sticky-table');

    if (this.table) {
      this.realTable = this.table.querySelector('#sticky-table-x-wrapper').firstChild;
      this.xWrapper = this.table.querySelector('#sticky-table-x-wrapper');
      this.yWrapper = this.table.querySelector('#sticky-table-y-wrapper');
      this.xScrollbar = this.table.querySelector('#x-scrollbar');
      this.yScrollbar = this.table.querySelector('#y-scrollbar');
      this.stickyHeader = this.table.querySelector('#sticky-header');
      this.stickyColumn = this.table.querySelector('#sticky-column');
      this.stickyCorner = this.table.querySelector('#sticky-corner');
      this.rightBottomCorner = this.table.querySelector('#right-bottom-corner');

      elementResizeEvent(this.realTable, this.onResize);
      elementResizeEvent(document.body, this.onResizeBody);

      this.xWrapper.addEventListener('scroll', this.onScrollX);

      this.onResize();
      setTimeout(this.onResize);
      this.addScrollBarEventHandlers();
    }
  }

  componentWillReceiveProps(nextProps) {
    const rc = getColRowCount(nextProps.children);

    this.setState({
      rowCount: rc.rowCount,
      columnCount: rc.columnCount
    });
  }

  componentDidUpdate() {
    this.onResize();
  }

  componentWillUnmount() {
    if (this.table) {
      this.xWrapper.removeEventListener('scroll', this.onScrollX);
    }
  }

  onResizeBody = () => {
    this.yWrapper.style.maxHeight = `${this.getNodeSize(document.getElementById('main')).height * 0.65}px`;
  }

  onResize = () => {
    this.setCornersVisibility();
    this.setRowHeights();
    this.setStickyCellsHeight();
    this.setColumnWidths();
    this.setScrollBarDims();
    this.setScrollBarWrapperDims();
  }

  onScrollX = () => {
    const scrollLeft = Math.max(this.xWrapper.scrollLeft, 0); // Can't have it being less than 0...
    this.stickyHeader.style.transform = `translate3d(${-1 * scrollLeft}px, 0, 0)`;
  }

  setCornersVisibility = () => {
    const tableWidth = this.table.offsetWidth;
    const tableHeight = this.table.offsetHeight;
    const contentWidth = this.xScrollbar.firstChild.offsetWidth;
    const contentHeight = this.yScrollbar.firstChild.offsetHeight;

    if (tableWidth >= contentWidth && tableHeight >= contentHeight) {
      this.rightBottomCorner.style.display = 'none';
      this.table.style.paddingBottom = '0';
      this.xWrapperMargin = 0;
      this.yWrapperMargin = 0;
    }

    if (tableWidth <= contentWidth && tableHeight <= contentHeight) {
      this.rightBottomCorner.style.display = 'block';
      this.table.style.paddingBottom = '13px';
      this.xWrapperMargin = 14;
      this.yWrapperMargin = 14;
    }

    if (tableWidth >= contentWidth && tableHeight <= contentHeight) {
      this.rightBottomCorner.style.display = 'none';
      this.table.style.paddingBottom = '0';
      this.xWrapperMargin = 0;
      this.yWrapperMargin = 0;
    }

    if (tableWidth <= contentWidth && tableHeight >= contentHeight) {
      this.rightBottomCorner.style.display = 'none';
      this.table.style.paddingBottom = '13px';
      this.xWrapperMargin = 0;
      this.yWrapperMargin = 0;
    }
  }

  setScrollBarDims = () => {
    this.xScrollbar.firstChild.style.width = `${this.getNodeSize(this.realTable.firstChild).width - this.stickyColumn.offsetWidth - this.xWrapperMargin}px`;
    this.yScrollbar.firstChild.style.height = `${this.getNodeSize(this.realTable).height - this.stickyHeader.offsetHeight - this.yWrapperMargin}px`;
  }

  setScrollBarWrapperDims = () => {
    this.xScrollbar.style.width = `calc(100% - ${this.stickyColumn.offsetWidth + this.xWrapperMargin}px)`;
    this.xScrollbar.style.left = `${this.stickyColumn.offsetWidth}px`;

    this.yScrollbar.style.height = `calc(100% - ${this.stickyHeader.offsetHeight + this.yWrapperMargin}px)`;
    this.yScrollbar.style.top = `${this.stickyHeader.offsetHeight}px`;
  }

  getNodeSize = (node) => {
    const nodeStyle = this.getStyle(node);
    const width = node.offsetWidth
      - parseFloat(nodeStyle.paddingLeft)
      - parseFloat(nodeStyle.paddingRight)
      - parseInt(nodeStyle.borderLeftWidth, 10)
      - parseInt(nodeStyle.borderRightWidth, 10);

    const height = node.offsetHeight
      - parseFloat(nodeStyle.paddingTop)
      - parseFloat(nodeStyle.paddingBottom)
      - parseInt(nodeStyle.borderTopWidth, 10)
      - parseInt(nodeStyle.borderBottomWidth, 10);

    return { width, height };
  }

  getStyle = (node) => {
    const browserSupportsComputedStyle = typeof getComputedStyle !== 'undefined';

    return browserSupportsComputedStyle ? getComputedStyle(node, null) : node.currentStyle;
  }

  getStickyCorner = (rows) => {
    const stickyCorner = [];
    const cell = React.Children.toArray(rows[0].props.children)[0];

    stickyCorner.push(
      <Row key="stickyCornerCell">
        {cell}
      </Row>
    );

    return stickyCorner;
  }

  getStickyHeader = (rows) => {
    const row = rows[0];
    const cells = [];

    React.Children.toArray(row.props.children).forEach((cell, c) => {
      cells.push(React.cloneElement(cell, { id: `sticky-header-cell-${c}`, key: `${c}stickyHeaderCol` }));
    });

    return (
      <Row id="sticky-header-row">
        {cells}
      </Row>
    );
  }

  getStickyColumn = (rows) => {
    let cells;
    const stickyRows = [];

    rows.forEach((row, r) => {
      cells = React.Children.toArray(row.props.children);

      stickyRows.push(
        <Row key={ `${r}stickyColRow` }>
          { cells[0] }
        </Row>
      );
    });

    return stickyRows;
  }

  setRowHeights = () => {
    let r;
    let cellToCopy;

    if (this.stickyColumnCount) {
      for (r = 0; r < this.rowCount; r += 1) {
        cellToCopy = this.realTable.childNodes[r].firstChild;

        if (cellToCopy) {
          const { height } = this.getNodeSize(cellToCopy);

          this.stickyColumn.firstChild.childNodes[r].firstChild.style.height = `${height}px`;

          if (r === 0 && this.stickyCorner.firstChild.firstChild) {
            this.stickyCorner.firstChild.firstChild.firstChild.style.height = `${height}px`;
          }
        }
      }
    }
  }

  setColumnWidths = () => {
    if (this.state.stickyHeaderCount) {
      for (let c = 0; c < this.state.columnCount; c += 1) {
        const cellToCopy = this.realTable.firstChild.childNodes[c];

        if (cellToCopy) {
          const { width } = this.getNodeSize(cellToCopy);
          let cell = this.table.querySelector(`#sticky-header-cell-${c}`);

          cell.style.width = `${width}px`;
          cell.style.minWidth = `${width}px`;
          cell.style.boxSizing = 'content-box';

          if (c === 0 && this.stickyCorner.firstChild.firstChild) {
            cell = this.stickyCorner.firstChild.firstChild.firstChild;

            cell.style.width = `${width}px`;
            cell.style.minWidth = `${width}px`;
          }
        }
      }
    }
  }

  setStickyCellsHeight = () => {
    const secondRow = this.realTable.firstChild.nextSibling;
    if (secondRow) {
      const rowHeight = secondRow.firstChild && secondRow.firstChild.offsetHeight;
      if (rowHeight) {
        for (let c = 0; c < this.state.rowCount; c += 1) {
          const cell = this.stickyColumn.firstChild.childNodes[c];

          if (c !== 0) { // skip header cell
            cell.style.height = `${rowHeight}px`;
          }
        }
      }
    }
  }

  addScrollBarEventHandlers = () => {
    // X Scrollbars
    this.xWrapper.addEventListener('scroll', this.scrollXScrollbar);
    this.xScrollbar.addEventListener('scroll', () => {
      if (!this.suppressScroll) {
        this.xWrapper.scrollLeft = this.xScrollbar.scrollLeft;
        this.suppressScroll = true;
      } else {
        this.suppressScroll = false;
      }
    });

    // Y Scrollbars
    this.yWrapper.addEventListener('scroll', this.scrollYScrollbar);
    this.yScrollbar.addEventListener('scroll', () => {
      if (!this.suppressScroll) {
        this.yWrapper.scrollTop = this.yScrollbar.scrollTop;
        this.suppressScroll = true;
      } else {
        this.suppressScroll = false;
      }
    });
  }

  scrollXScrollbar = () => {
    if (!this.suppressScroll) {
      this.xScrollbar.scrollLeft = this.xWrapper.scrollLeft;
      this.suppressScroll = true;
    } else {
      this.suppressScroll = false;
    }
  }

  scrollYScrollbar = () => {
    if (!this.suppressScroll) {
      this.yScrollbar.scrollTop = this.yWrapper.scrollTop;
      this.suppressScroll = true;
    } else {
      this.suppressScroll = false;
    }
  }

  render() {
    const { children } = this.props;
    const { stickyColumnCount, stickyHeaderCount } = this.state;
    const rows = React.Children.toArray(children);

    let stickyColumn;
    let stickyHeader;
    let stickyCorner;

    if (rows.length) {
      if (stickyColumnCount > 0 && stickyHeaderCount > 0) {
        stickyCorner = this.getStickyCorner(rows);
      }
      if (stickyColumnCount > 0) {
        stickyColumn = this.getStickyColumn(rows);
      }
      if (stickyHeaderCount > 0) {
        stickyHeader = this.getStickyHeader(rows);
      }
    }

    return (
      <div className={ styles.stickyTable } id="frood-sticky-table">
        <div className={ styles.xScrollbar } id="x-scrollbar"><div /></div>
        <div className={ styles.rightBottomCorner } id="right-bottom-corner" />
        <div className={ styles.leftBottomCorner } id="left-bottom-corner" />
        <div className={ styles.yScrollbar } id="y-scrollbar"><div /></div>
        <div className={ styles.stickyCorner } id="sticky-corner">
          <Table>{stickyCorner}</Table>
        </div>
        <div className={ styles.stickyHeader } id="sticky-header">
          <Table>{stickyHeader}</Table>
        </div>
        <div className={ styles.stickyTableYWrapper } style={ { maxHeight: this.maxHeight } } id="sticky-table-y-wrapper">
          <div className={ styles.stickyColumn } id="sticky-column">
            <Table>{stickyColumn}</Table>
          </div>
          <div className={ styles.stickyTableXWrapper } id="sticky-table-x-wrapper">
            <Table>{rows}</Table>
          </div>
        </div>
      </div>
    );
  }
}

StickyTable.propTypes = {
  stickyColumnCount: PropTypes.number,
  stickyHeaderCount: PropTypes.number,
  children: PropTypes.array,
};

export default StickyTable;

