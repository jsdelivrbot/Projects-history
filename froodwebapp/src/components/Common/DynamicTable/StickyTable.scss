@import '~styles/colors.scss';
@import '~styles/fonts.scss';

$cell-padding: 0.5rem 0.75rem;
$cursor: pointer;

.dynamicTable {
  position: relative;
  border-bottom: 1px solid #d9d9d9;
  border-left: 1px solid #d9d9d9;
  border-right: 1px solid #d9d9d9;
  border-top: none;
}

.actionText {
  text-decoration: underline;
  color: $main-black;
}

.dynamicTableAlert {
  padding: 0.5rem;
  text-align: center;
  border-radius: 0;
  border: none;
  background-color: $main-white;
}

.stickyTable {
  position: relative;
  overflow: hidden;
  height: 100%;
  .cell {
    display: table-cell;
    white-space: pre;
    border-bottom: 1px solid $main-gray;
    border-collapse: collapse;
    font-weight: $table-body-font-weight;
    padding: $cell-padding;
  }
  .row {
    display: table-row;
    cursor: $cursor;
    &:hover {
      background-color: $row-hover-color;
    }
  }

  .rightBottomCorner {
    position: absolute;
    height: 0.95rem;
    width: 0.95rem;
    bottom: -1px;
    right: -1px;
    background-color: #fafafa;
    z-index: 20;
  }

  .table{
    display: table;
  }
  .stickyCorner {
    position: absolute;
    left: 0;
    top: 0;
    z-index: 3;
    .cell {
      background-color: $main-gray;
      border-bottom: none;
    }
  }
  .stickyHeader {
    position: absolute;
    z-index: 2;
    div[class*="row"] {
      div[class*="cell"]:last-child {
        padding-right: 2rem;
      }
    }
    .cell {
      background-color: $main-gray;
      color: rgba(0,0,0,.85);
      font-weight: 200;
      text-transform: uppercase;
    }
  }
  .stickyColumn {
    position: absolute;
    z-index: 2;
    background-color: #fff;
    border-right: 1px solid $main-gray;
    .table {
      .row:first-child {
        .cell {
          border-bottom: none;
        }
      }
    }
    .cell {
      vertical-align: middle;
    }
  }
  .stickyTableXWrapper {
    z-index: 0;
    position: relative;
    width: 100%;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
    .table {
      /*
      Hack to make table paint on it's own layer.
      Because of this it won't paint when the scrollbar paints.
      This one was crucial for performance.
      Scrolling is probably 10-20x smoother.
      */
      width: 100%;
      transform: translateZ(0);
      -webkit-transform: translateZ(0);
      div[class*="row"]:first-child {
        background-color: $main-gray;
        color: rgba(0,0,0,.85);
        font-weight: $table-header-font-weight;
        text-transform: uppercase;
        div[class*="cell"] {
          border-bottom: none;
        }
      }
      div[class*="row"] {
        div[class*="cell"]:last-child {
          padding-right: 2rem;
        }
      }
    }
  }
  .stickyTableYWrapper {
    z-index: 0;
    position: relative;
    height: 100%;
    -webkit-overflow-scrolling: touch;
    overflow-y: auto;
    /* Hack to hide scrollbars */
    padding-right: 17px;
    margin-right: -17px;
  }
}

.xScrollbar {
  position: absolute;
  background-color: transparent;
  z-index: 5;
  overflow: auto;
  bottom: -1px;
  height: 20px;
  transform: translateZ(0);
  -webkit-transform: translateZ(0);
  div {
    height: 100%;
  }
}

.yScrollbar {
  position: absolute;
  background-color: transparent;
  z-index: 5;
  overflow: auto;
  right: -1px;
  top: 0px;
  height: 100%;
  width: 20px;
  transform: translateZ(0);
  -webkit-transform: translateZ(0);
  div {
    width: 100%;
  }
}

/* Scrollbars */
.stickyTableYWrapper::-webkit-scrollbar,
.stickyTableXWrapper::-webkit-scrollbar {
  display: none;
}

