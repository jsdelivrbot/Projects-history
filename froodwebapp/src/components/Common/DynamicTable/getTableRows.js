/* eslint-disable babel/new-cap, react/no-array-index-key, react/jsx-filename-extension */
import React from 'react';
import { Checkbox } from 'antd';
import { Row, Cell } from './TableComponents';
import ConnectedDraggableHeader from './DraggableHeader';
import styles from './StickyTable.scss';

const getTableRows = (
  headers,
  data,
  selectedRows,
  handleRowClick,
  handleDownloadItem,
  handleSelectRows,
  handleSort,
  actionColumnVisible,
  moveHeader,
  saveHeaderOrder
) => {
  const rows = [];
  rows.push(
    <Row key="headers">
      <Cell>
        <Checkbox
          onClick={ e => handleSelectRows('all', e.target.checked) }
        />
      </Cell>
      { headers.map((header, headerIndex) => (
        <ConnectedDraggableHeader
          key={ header.key }
          index={ headerIndex }
          moveHeader={ moveHeader }
          saveHeaderOrder={ saveHeaderOrder }
          isSortable={ header.isSortable }
          handleSort={ handleSort }
          headerId={ header.id }
          headerText={ header.name }
          headerDataType={ header.dataType }
        />
      ))}
      { actionColumnVisible &&
        <Cell>
          Action
        </Cell>
      }
    </Row>
  );

  data.map((item, rowIndex) => (
    rows.push(
      <Row id={ item.id } key={ rowIndex } onClick={ handleRowClick }>
        <Cell>
          <Checkbox
            onClick={ () => handleSelectRows(item.id) }
            checked={ selectedRows.indexOf(item.id) !== -1 }
          />
        </Cell>
        {
          headers.map((header, headerIndex) => (
            <Cell
              key={ `${rowIndex}-${headerIndex}` }
              style={ { textAlign: header.dataType.includes('decimal') ? 'right' : 'left' } }
            >
              { (!header.isEditable
               && header.dataType.includes('decimal')
               && Number(item[header.dataAlias]).toFixed(2))
               || item[header.dataAlias]
              }
            </Cell>
          ))
        }
        { actionColumnVisible &&
          <Cell>
            <span>
              <a
                role="button"
                tabIndex="-1"
                id={ item.id }
                className={ styles.actionText }
                onClick={ handleDownloadItem }
              >
                Download
              </a>
            </span>
          </Cell>
        }
      </Row>
    )
  ));
  return rows;
};

export default getTableRows;
