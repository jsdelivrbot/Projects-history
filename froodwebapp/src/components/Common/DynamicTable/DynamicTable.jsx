/* eslint-disable babel/new-cap, react/no-array-index-key, react/no-unused-prop-types */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Spin, Alert } from 'antd';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import update from 'immutability-helper';
import getTableRows from './getTableRows';
import StickyTable from './StickyTable';
import BulkMenu from './BulkMenu/BulkMenu';
import styles from './StickyTable.scss';

export class DynamicTable extends Component {
  constructor(props) {
    super(props);
    this.state = {
      headers: props.headers,
      data: props.data,
      selectedRows: props.selectedRows,
      tableRows: getTableRows(
        props.headers,
        props.data,
        props.selectedRows,
        props.handleRowClick,
        props.handleDownloadItem,
        props.handleSelectRows,
        props.handleSort,
        props.actionColumnVisible,
        this.moveHeader,
        this.saveHeaderOrder
      )
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.loadingTableData) {
      this.setState({
        headers: nextProps.headers,
        data: nextProps.data,
        selectedRows: nextProps.selectedRows,
        tableRows: getTableRows(
          nextProps.headers,
          nextProps.data,
          nextProps.selectedRows,
          nextProps.handleRowClick,
          nextProps.handleDownloadItem,
          nextProps.handleSelectRows,
          nextProps.handleSort,
          nextProps.actionColumnVisible,
          this.moveHeader,
          this.saveHeaderOrder
        )
      });
    }
  }

  moveHeader = (dragIndex, hoverIndex) => {
    const { headers } = this.state;
    const dragHeader = headers[dragIndex];

    const newState = update(this.state, {
      headers: {
        $splice: [
          [dragIndex, 1],
          [hoverIndex, 0, dragHeader],
        ],
      }
    });

    this.setState({
      headers: newState.headers,
      tableRows: getTableRows(
        newState.headers,
        this.state.data,
        this.state.selectedRows,
        this.props.handleRowClick,
        this.props.handleDownloadItem,
        this.props.handleSelectRows,
        this.props.handleSort,
        this.props.actionColumnVisible,
        this.moveHeader,
        this.saveHeaderOrder
      )
    });
  }

  saveHeaderOrder = () => {
    const { headers } = this.state;

    this.props.handleUpdateColumnsOrder(headers.map((header, index) => ({
      id: header.id,
      order: index
    })));
  }

  render() {
    const { tableRows, selectedRows } = this.state;
    const { loadingTableData } = this.props;

    return (
      <div id="dynamicTable" className={ styles.dynamicTable }>
        { (tableRows.length === 1 &&
          !loadingTableData &&
          <Alert
            message="No Data"
            type="info"
            className={ styles.dynamicTableAlert }
          />) ||
          <div>
            { selectedRows.length !== 0 &&
              <BulkMenu
                selectedRowsNumber={ selectedRows.length }
              />
            }
            <Spin
              tip="Loading..."
              spinning={ loadingTableData }
            >
              <StickyTable>
                { tableRows }
              </StickyTable>
            </Spin>
          </div>
        }
      </div>
    );
  }
}

DynamicTable.propTypes = {
  data: PropTypes.array,
  headers: PropTypes.array,
  selectedRows: PropTypes.array,
  loadingTableData: PropTypes.bool,
  actionColumnVisible: PropTypes.bool,
  handleSelectRows: PropTypes.func.isRequired,
  handleSort: PropTypes.func.isRequired,
  handleRowClick: PropTypes.func,
  handleUpdateColumnsOrder: PropTypes.func.isRequired,
  handleDownloadItem: PropTypes.func
};

export default DragDropContext(HTML5Backend)(DynamicTable);
