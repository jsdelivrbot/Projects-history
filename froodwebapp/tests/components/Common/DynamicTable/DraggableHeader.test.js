import React from 'react';
import { DraggableHeader } from 'components/Common/DynamicTable/DraggableHeader';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, it, expect } = global;

describe('DraggableHeader', () => {
  it('capturing Snapshot of DraggableHeader', () => {
    const props = {
      headerId: 1,
      headerText: 'Some text',
      headerDataType: 'string',
      index: 1,
      isDragging: false,
      isSortable: true,
      handleSort: jest.fn(),
      saveHeaderOrder: jest.fn(),
      moveHeader: jest.fn(),
      connectDragSource(dragSource) {
        return dragSource;
      },
      connectDropTarget(dropTarget) {
        return dropTarget;
      }
    };
    const component = shallow(<DraggableHeader { ...props } />);
    const view = toJson(component);
    const actual = expect(view);
    actual.toMatchSnapshot();
  });
});
