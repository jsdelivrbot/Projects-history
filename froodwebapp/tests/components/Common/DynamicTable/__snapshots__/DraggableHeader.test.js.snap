// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableHeader capturing Snapshot of DraggableHeader 1`] = `
<div
  className="cell"
  id="sticky-header-cell-2"
  key="Some text"
  style={
    Object {
      "cursor": "move",
      "opacity": 1,
      "textAlign": "left",
    }
  }
>
  Some text
</div>
`;
